src/FrameDump.cpp

```diff
@@ -0,0 +1,209 @@
+    memcpy(TexMtx, GPU.GPU3D.TexMatrix, sizeof(GPU.GPU3D.TexMatrix));
+    MatrixMode = GPU.GPU3D.MatrixMode;
+    Polygon = GPU.GPU3D.PolygonMode;
+    VtxXY = GPU.GPU3D.CurVertex[1] << 16 | GPU.GPU3D.CurVertex[0];
+    VtxZ = GPU.GPU3D.CurVertex[2];
+    TexCoord = GPU.GPU3D.TexCoords[1] << 16 | GPU.GPU3D.TexCoords[0]; // use final texcoords so i dont have to worry about setting up the matrix & params properly.
+    TexParam = GPU.GPU3D.TexParam;
+    TexPalette = GPU.GPU3D.TexPalette;
+    DiffAmbi = GPU.GPU3D.MatAmbient[2] << 26 | GPU.GPU3D.MatAmbient[1] << 21 | GPU.GPU3D.MatAmbient[0] << 16
+             | GPU.GPU3D.MatDiffuse[2] << 10 | GPU.GPU3D.MatDiffuse[1] << 5 | GPU.GPU3D.MatDiffuse[0];
+    SpecEmis = GPU.GPU3D.MatEmission[2] << 26 | GPU.GPU3D.MatEmission[1] << 21 | GPU.GPU3D.MatEmission[0] << 16
+             | GPU.GPU3D.MatSpecular[2] << 10 | GPU.GPU3D.MatSpecular[1] << 5 | GPU.GPU3D.MatSpecular[0];
+    for (int i = 0; i < 32; i++)
+    {
+        u8 tbl = i << 2;
+        Shininess[i] = GPU.GPU3D.ShininessTable[tbl];
+        Shininess[i] |= GPU.GPU3D.ShininessTable[tbl + 1] << 8;
+        Shininess[i] |= GPU.GPU3D.ShininessTable[tbl + 2] << 16;
+        Shininess[i] |= GPU.GPU3D.ShininessTable[tbl + 3] << 24;
+    }
+    for (u32 i = 0; i < 4; i++)
+    {
+        LightVec[i] = i << 30 | ((u32)GPU.GPU3D.LightDirection[i][2] & 0x3FF) << 20 | ((u32)GPU.GPU3D.LightDirection[i][1] & 0x3FF) << 10 | ((u32)GPU.GPU3D.LightDirection[i][0] & 0x3FF);
+        LightColor[i] = i << 30 | GPU.GPU3D.LightColor[i][2] << 10 | GPU.GPU3D.LightColor[i][1] << 5 | GPU.GPU3D.LightColor[i][0];
+    }
+    SwapBuffer = GPU.GPU3D.FlushAttributes;
+
+    // they should already be empty, but just in case.
+    Cmds.clear();
+    Params.clear();
+}
+
+bool FrameDump::FinFrameDump()
+{
+    std::string filename = ROMManager::GetFrameDumpName();
+    Platform::FileHandle* file = Platform::OpenLocalFile(filename, Platform::FileMode::Write);
+    
+    if (file == nullptr) return true;
+
+    // save final state vars to file
+    Platform::FileWrite(&GPU.GPU3D.RenderDispCnt, 1, 2, file); // only write two bytes, (only 16 bits are used, and thus only those are stored for frame dumps)
+    Platform::FileWrite(GPU.GPU3D.RenderEdgeTable, 1, sizeof(GPU.GPU3D.RenderEdgeTable), file);
+    Platform::FileWrite(&GPU.GPU3D.RenderAlphaRef, 1, sizeof(GPU.GPU3D.RenderAlphaRef), file);
+    Platform::FileWrite(&GPU.GPU3D.RenderClearAttr1, 1, sizeof(GPU.GPU3D.RenderClearAttr1), file);
+    Platform::FileWrite(&GPU.GPU3D.RenderClearAttr2, 1, sizeof(GPU.GPU3D.RenderClearAttr2), file);
+    Platform::FileWrite(&GPU.GPU3D.RenderFogColor, 1, sizeof(GPU.GPU3D.RenderFogColor), file);
+    Platform::FileWrite(&GPU.GPU3D.RenderFogOffset, 1, 2, file); // only write two bytes
+    Platform::FileWrite(&GPU.GPU3D.RenderFogDensityTable[1], 1, sizeof(GPU.GPU3D.RenderFogDensityTable[0])*32, file); // only write the 32 "real" entries of the density table
```

Comment:
Did you mean to start from `RenderFogDensityTable[0]`?

Reply:
nope, entries 0 and 33 are duplicates of 1 and 32

Reply:
I'd note that fact in a comment, then, just to ensure that someone trying to be helpful in the future doesn't undo it. My rule of thumb for comments; if it looks weird, comment it. Starting anything from index 1 looks weird.